Spmc queue

bartoszmodelski opened the pull request:
SPMC Queue. Similar design to MPMC Queue but owner of the queue (the single producer) can be far more relaxed. 

Includes: 
- Interface matching Domainslib 
- Resizing
- Tests and benchmarks

lyrm reviewed (commented):
General note : this is just a review of the implementation. When the current issues are fixed, I will review the other files.

A lot of good idea in this queue, but also some issues for which I don't see an immediate solution. 

In addition of  the code review, I have this test, that may loop forever. Not sure it is related to the found issues. The test is just a domain trying to `resize` a queue when an other one tries to `steal` from it. It does not always loop. 
```ocaml
let test () =
  let size_exp = 4 in
  let queue = create ~size_exponent:size_exp () in
  let size = 1 lsl size_exp in
  let sema = Semaphore.Binary.make false in

  for i = 0 to size / 2 do
    Local.push queue i |> ignore
  done;
  let d1 =
    Domain.spawn (fun () ->
        while not (Semaphore.Binary.try_acquire sema) do
          Domain.cpu_relax ()
        done;
        Local.resize queue)
  in
  let d2 =
    Domain.spawn (fun () ->
        let q' = create ~size_exponent:size_exp () in
        Semaphore.Binary.release sema;        
        Local.steal ~from:queue q' |> ignore;
        q')
  in
  Domain.join d1;
  let _q' = Domain.join d2 in
  ()
```

bartoszmodelski commented on src/spmc_queue.ml:
```diff
@@ -0,0 +1,244 @@
+    else
+      let index = tail_val land mask in
+      let cell = Array.get array index in
+      while Option.is_some (Atomic.get cell) do
+        ()
+      done;
```
I think this is the most practical approach, scheduler will try to push it into local runqueue and if it cannot fit there, the task will go into overflow queue that's unbounded. This busy waiting is here because from performance perspective it probably doesn't make sense to return at this point.

We can add a formally lockfree method that spins to qualify for the repo and let user choose. There's probably also a window to make this lock-free without spinning but showing linearization will be a non-trivial lift.

bartoszmodelski commented on src/spmc_queue.ml:
```diff
@@ -0,0 +1,244 @@
+
+  let steal ~from ({ mask = local_mask; _ } as local) =
+    (* need to be careful with from queue, which is not local *)
+    let local_size = Atomic.get local_mask + 1 in
```
Steal generally works on the assumption that there is space in the queue (that can be ensured with other functions). If we want to support more than that, we would have to scan the queue to ensure physical space as well, but both seem easy enough to add.

bartoszmodelski commented on src/spmc_queue.ml:
```diff
@@ -0,0 +1,244 @@
+          do
+            value := Atomic.get cell
+          done;
+          let value_exn = function None -> assert false | Some v -> v in
```
Yeah, this is to extract the `t`, so it can be pushed into the local queue below.

bartoszmodelski commented on src/spmc_queue.ml:
```diff
@@ -0,0 +1,244 @@
+          let from_index = (from_head_val + i) land from_mask_val in
+          let cell = Array.get from_array from_index in
+          let value = ref (Atomic.get cell) in
+          while
+            not
+              (Option.is_some !value && Atomic.compare_and_set cell !value None)
+          do
+            value := Atomic.get cell
+          done;
```
Agreed, can be changed to assert.

bartoszmodelski commented on src/spmc_queue.ml:
```diff
@@ -0,0 +1,244 @@
+            value := Atomic.get cell
+          done;
+          let value_exn = function None -> assert false | Some v -> v in
+          while not (push local (value_exn !value)) do
+            ()
+          done
```
As in the other case, `stealer` has to ensure there's physical space in the queue. If it does not have such certainty, it needs to call `is_empty` before trying to steal.

We could add the index checks and array scan here, and make `steal` return `Not_enough_space if they fail. I'm weakly hesitant, since it's not the usual use-case, but it should fine to include.

bartoszmodelski commented:
Thanks for the test, I can reproduce it.

bartoszmodelski commented on src/spmc_queue.ml:
```diff
@@ -0,0 +1,244 @@
+    (* just for extra safety, wait until stealers finish - should not
+       be necessary for correctness *)
+    let num_of_items_to_ensure_taken = temp_head_val - tail_val in
+    for i = 0 to num_of_items_to_ensure_taken - 1 do
+      let index = (tail_val + i) land mask_val in
+      while Option.is_some (Atomic.get (Array.get array_val index)) do
+        ()
+      done
+    done;
```
This inner loop cannot be an assert, as stealers might still be removing items from the queue. 

But the conclusion is correct; both loops are not required for correctness. It's more for tractability of `steal`. Otherwise there's many ways to introduce subtle bugs, that we have no way of preventing, that is any access to mask, array, tail after CAS on head will break things.

bartoszmodelski commented:
@lyrm I've added the test you posted to the suite and patched resizing code. 

The issue stemmed from `resize` happening between `steal`'s taking `array` and doing CAS on `head`. It was probably fixable with reordering and extra guards but it's a bit of rabbit hole; for model checker to pass we'd have to allow N successful resizing in the middle of a steal. 

So I've taken a different approach; instead of having `steal` and `resize` implemented in the core structure (what requires precise ordering), `resize` is now implemented as a wrapper around bounded spmc queue.

lyrm commented on src/spmc_queue.ml:
```diff
@@ -232,13 +145,69 @@ module M = struct
+        let value = Atomic.get cell in
+        assert (Atomic.compare_and_set cell value None);
+        assert (Option.is_some value);
+        match value with None -> assert false | Some v -> v
```
Two last lines  could be replaced by `Option.get value`

lyrm commented on src/spmc_queue.ml:
```diff
@@ -0,0 +1,244 @@
+      let index = old_head_val land mask in
+      let cell = Array.get array index in
+      let item = ref (Atomic.get cell) in
+      while
+        not (Option.is_some !item && Atomic.compare_and_set cell !item None)
+      do
+        item := Atomic.get cell
+      done;
```
Thanks for the clarification about optimistic !

Can this loop be removed and replace by an assert ?

bartoszmodelski replied:
I think so, yes

lyrm commented:
I am worked on a solution for the lock-freedom issue of `push` that use an help mechanism : if the producer encounters a non-empty cell (meaning a stealer has not finished its works), it will try to complete the operation itself. I am still working on it but you can have a look [here](https://github.com/lyrm/lockfree/tree/spmc_queue_with_help). 

Performances for one single stealer seem pretty similar than before (see [tests to launch at top level](https://github.com/lyrm/lockfree/blob/spmc_queue_with_help/test/test_spmc.ml)) but I obviously need to check they are not decreasing a lot with multiple stealers. 

Also, this implementation is not passing some of your tests and I need to check why.

lyrm commented:
Unfortunately, the helping mechanism is not going to work the way I did it.  If anyone else has an idea to correct the lock freedom issue of the `push` function, please do !

bartoszmodelski commented:
As discussed, we could probably offer a fully lockfree implementation by simplifying `steal` to steal-one (and have it scan array rather than increment `tail`). Such a twist would conform to the interface accepted by domainslib (`ws_deque`). But I'll wait for future reviewer's input.

bartoszmodelski commented:
I'll resubmit a simplified version of this at some point.
